ballerina/websocket_service.bal

```diff
@@ -24,70 +25,108 @@ isolated service class WsService {
+                                        ? wsText.cloneWithType(WSPayload) : value:fromJsonString(text);
+        if wsPayload is error {
+            json payload = {errors: [{message: "Invalid format in WebSocket payload"}]};
+            check sendWebSocketResponse(caller, self.customHeaders, WS_ERROR, payload);
+            closeConnection(caller);
+            return;
+        }
+        string wsType = wsPayload is WSPayload ? <string>wsPayload.'type : DEFAULT_VALUE;
+        string connectionId = wsPayload is WSPayload && wsPayload?.id !is () ? <string>wsPayload?.id : DEFAULT_VALUE;
+
+        if wsType == WS_INIT {
+            lock {
+                if self.initiatedConnection {
+                    closeConnection(caller, 4429, "Too many initialisation requests");
+                    return;
+                }
+                self.initiatedConnection = true;
+            }
+            check caller->writeMessage({"type": WS_ACK});
+        } else if wsType == WS_SUBSCRIBE || wsType == WS_START || !self.customHeaders.hasKey(WS_SUB_PROTOCOL) {
+            lock {
+                if self.customHeaders.hasKey(WS_SUB_PROTOCOL) {
+                    if !self.initiatedConnection {
+                        closeConnection(caller, 4401, "Unauthorized");
+                        return;
+                    }
+                    if self.activeConnections.indexOf(connectionId) !is () {
+                        closeConnection(caller, 4409, string `Subscriber for ${connectionId} already exists`);
+                        return;
+                    }
+                    string[] connections = [connectionId];
+                    foreach string i in self.activeConnections {
+                        connections.push(i);
                     }
-                    next = sourceStream.iterator().next();
+                    self.activeConnections = connections.cloneReadOnly();
                 }
+            }
+            parser:OperationNode|json node = validateSubscriptionPayload(wsPayload, self.engine);
+            if node is parser:OperationNode {
+                check executeOperation(self.engine, self.context, self.schema, self.customHeaders, caller,
+                                       connectionId, node);
             } else {
-                check caller->writeTextMessage(sourceStream.toJsonString());
+                check sendWebSocketResponse(caller, self.customHeaders, WS_ERROR, node, connectionId);
+                closeConnection(caller);
+            }
+        } else if wsType == WS_STOP || wsType == WS_COMPLETE {
+            lock {
+                string[] connections = [];
+                foreach string i in self.activeConnections {
+                    connections.push(i);
+                }
+                int? index = array:indexOf(connections, connectionId);
+                if index is int {
+                    _ = connections.remove(index);
+                }
```

Comment:
What's the reason here to clone the `self.activeConnections` to a local variable and re-assigning it back?

Reply:
`self.activeConnections` should be immutable here. Therefore I used `readonly` when the variable is initiated. But here I used a local variable to use the array as a mutable object.

Reply:
Why does it need to be immutable?